Consolide 'What is a feature flag' docs (#8050)
Unleash/unleash: website/docs/topics/concepts/what-is-a-feature-flag.md

---
title: What is a feature flag and why are feature flags used?
slug: /what-is-a-feature-flag
description: Feature flags let you control software features in real time, enabling safer deployments, better testing, and faster innovation.
---

import Figure from '@site/src/components/Figure/Figure.tsx'

Feature flags allow you to release, test, and manage features and functionality across your application without changing the source code. Organizations use added control and flexibility to deliver more and higher quality features with reduced cost, time, and risk.

In this guide, we’ll introduce you to the basics of feature flags and why so many development teams choose them.

## Why developers use feature flags

> In 2014, [Ivar Østhus](https://www.linkedin.com/in/ivarconr/overlay/about-this-profile/), our CTO, was working as a software developer at FINN.no, Norway’s largest online marketplace. Getting new features into production was a pain, so, as a developer, he looked around and wondered if he could write some code to make it easier. His initial goal was to protect features under development, allowing his team to transition to [trunk-based development](https://www.getunleash.io/blog/how-to-implement-trunk-based-development-a-practical-guide)—a practice recommended by the [State of DevOps Report](https://cloud.google.com/devops/state-of-devops).

Reviewer: I like that idea and agree that 1st person is more impactful here.  Does it need to be collapsible?  I think we can just switch back to 1st person, no?

> In 2014, [Ivar Østhus](https://www.linkedin.com/in/ivarconr/overlay/about-this-profile/), our CTO, was working as a software developer at FINN.no, Norway’s largest online marketplace. Getting new features into production was a pain, so, as a developer, he looked around and wondered if he could write some code to make it easier. His initial goal was to protect features under development, allowing his team to transition to [trunk-based development](https://www.getunleash.io/blog/how-to-implement-trunk-based-development-a-practical-guide)—a practice recommended by the [State of DevOps Report](https://cloud.google.com/devops/state-of-devops).

Author: I think it would be a good idea to go for something different visually than the rest of the text. Just to help the reader switch between the different narratives. I went for this option:

https://unleash-docs-git-melinda-what-is-a-feature-flag-unleash-team.vercel.app/what-is-a-feature-flag#why-developers-use-feature-flags

wdyt?

Frustrated by the slow pace and the additional complexity of feature branches, Ivar explored ways to streamline it through code and ended up creating [Unleash](https://github.com/Unleash/unleash), a popular open-source feature flag solution that solves these problems by decoupling code deployments and feature releases.

## What is a feature flag?

A feature flag is a software engineering technique that turns features or functionality on or off without modifying the source code or requiring a redeploy. It’s also referred to as feature toggles, switches, flippers, or bits.

Feature flags enable you to release and test new features by making them available to a specific group of users, or no users at all, with the power to immediately turn it off without any risk to the rest of your application.

The control that feature flags allow means that they’re already becoming a popular superpower in feature lifecycle management. Combined with feature management platforms and processes, they can enable a cultural shift within organizations towards more agile and experimental development and ways of serving users.

> ![Diagram demonstrating the relationship between a CI/CD pipeline, an application, and its feature flags](/img/how-feature-flags-work-diagram.png).

Reviewer: Ok, would love your thoughts on standardizing docs diagrams.

> ![Diagram demonstrating the relationship between a CI/CD pipeline, an application, and its feature flags](/img/how-feature-flags-work-diagram.png).

Author: For now, I just redrew it in a vertical format, this helps a lot already:

https://unleash-docs-git-melinda-what-is-a-feature-flag-unleash-team.vercel.app/what-is-a-feature-flag#why-developers-use-feature-flags

### Feature management​

Feature management involves the strategic use of feature flags to control the lifecycle and rollout of features. This includes scheduling feature releases, managing access based on user roles, and monitoring feature performance and usage. Effective feature management ensures that new features are released in a controlled and measured manner, reducing risks and improving overall product stability. 

Additionally, feature flags can be implemented in a secure and scalable manner to support enterprise-level requirements. By using robust feature flagging systems, organizations can ensure that their feature management processes comply with security standards and can scale to handle large user bases without compromising performance.

### Compliance with FedRAMP, SOC2, ISO27001, Air-gap, and others

> Feature flags can be managed in environments that require strict compliance with standards such as FedRAMP, SOC2, ISO27001, and air-gap configurations. By ensuring that feature flagging systems meet these regulatory requirements, organizations can maintain compliance while benefiting from agile feature management.

Reviewer: A use case for Unleash in particular is FedRAMP compliance.  Other vendors on the market can't do this because they can't be run in a customers compliant environment.  So in this case, it is more of an Unleash use case, than a general ff use case, if that makes sense.

## Applying feature flag best practices with Unleash

> The risk and cost of building software the old way are too high. When developers are forced to use tools and processes they hate, like infrequent high-stakes releases based on complex branching strategies, innovation is reduced by 50%. Additionally, when code is deployed and released in this manner, rolling back after discovering an issue can be difficult and time-consuming. On average, application downtime costs organizations $400,000 per hour, and more than half of organizations face unexpected downtime every two months. 

Reviewer: Yeah, the 50% number comes from this study: https://www.microsoft.com/en-us/research/publication/devex-in-action-a-study-of-its-tangible-impacts/

And the $400,000 is from an IBM study we site here: https://www.getunleash.io/blog/how-feature-management-reduces-downtime-risk

> The risk and cost of building software the old way are too high. When developers are forced to use tools and processes they hate, like infrequent high-stakes releases based on complex branching strategies, innovation is reduced by 50%. Additionally, when code is deployed and released in this manner, rolling back after discovering an issue can be difficult and time-consuming. On average, application downtime costs organizations $400,000 per hour, and more than half of organizations face unexpected downtime every two months. 

Reviewer: I think we should keep the numbers and link

> The risk and cost of building software the old way are too high. When developers are forced to use tools and processes they hate, like infrequent high-stakes releases based on complex branching strategies, innovation is reduced by 50%. Additionally, when code is deployed and released in this manner, rolling back after discovering an issue can be difficult and time-consuming. On average, application downtime costs organizations $400,000 per hour, and more than half of organizations face unexpected downtime every two months. 

Author: Added the links, thanks!

> Unleash is on a mission to make developers’ lives easier. Individual developers love Unleash because it removes the pain of testing and deploying new features so they have more time and energy to innovate. Unleash is trusted by thousands of companies in production including Visa, Wayfair, Lloyd’s Banking Group, and Samsung. While we serve the needs of some of the world’s largest and most security-conscious organizations, we are also rated the *Easiest to Use in Feature Management software* by G2. 

Reviewer: Those rating jump around a bit from quarter to quarter. We are regularly at the top.  We should keep and adjust if that changes considerably.

If you want to learn more about how to implement feature flags at scale, check out the following resources:
- [Feature Flag Tutorials](/feature-flag-tutorials)
- [Best practices for using feature flags at scale](../feature-flags/best-practices-using-feature-flags-at-scale.md)
- [Best practices for building and scaling feature flags](../feature-flags/feature-flag-best-practices.md)

> - [Try Unleash for Free](https://www.getunleash.io/pricing)

Reviewer: Yes, so they can see all the options.  They might want to try the hosted offering with a free 14 day trial instead of spinning up their own Unleash instance.

---
title: What is a feature flag and why are feature flags used?
slug: /what-is-a-feature-flag
description: Feature flags let you control software features in real time, enabling safer deployments, better testing, and faster innovation.
---

import Figure from '@site/src/components/Figure/Figure.tsx'

Feature flags allow you to release, test, and manage features and functionality across your application without changing the source code. Organizations use added control and flexibility to deliver more and higher quality features with reduced cost, time, and risk.

In this guide, we’ll introduce you to the basics of feature flags and why so many development teams choose them.

## Why developers use feature flags

> *The story of how Unleash began by [Ivar Østhus](https://www.linkedin.com/in/ivarconr/overlay/about-this-profile/)*

Reviewer: ```suggestion
*The easiest way to explain why developers use feature flags is to understand the story of why [Ivar Østhus](https://github.com/ivarconr), 
 our CTO, created the Unleash feature flag service in the first place. Here's Ivar in his own words:*
```

Contrary to the short-lived nature of most feature flags, kill switches are inherently long-term or even permanent tools within a system’s architecture. Their enduring presence allows for ongoing protection against unforeseen critical events. However, these switches must be regularly reviewed and tested to ensure they remain functional and relevant, adapting to the evolving system and operational requirements.

### Operational flags

- **Purpose**: Transition between technical implementations with minimal risk.
- **Benefits**: Thorough validation of changes in a live environment without disrupting the user experience.
- **Lifespan**: Short-lived; a few days to a couple of weeks.

> Operational flags manage a system’s technical aspects that do not directly influence user-visible features but are crucial for the application’s underlying stability and performance. For example, transitioning from one library version to another, where operational flags can toggle between the old and new implementations to ensure the update does not adversely affect the system. By enabling real-time switching between different operational states or configurations, these flags allow for thorough validation of changes in a live environment without disrupting the user experience.

Author: ```suggestion
Operational flags manage a system’s technical aspects that do not directly influence user-visible features but are crucial for the application’s underlying stability and performance. For example, during a library version upgrade, operational flags can toggle between the old and new implementations to ensure the update does not adversely affect the system. By enabling real-time switching between different operational states or configurations, these flags allow for thorough validation of changes in a live environment without disrupting the user experience.
```
